### DISCUSSION TITLE
(Proposal) Part Hiding Options

### DISCUSSION THREAD

**Discussion by eriese** [freesewing/freesewing#3586] [Forum]
# Reasoning
We're running into confusion about which options hide which types of dependencies, and how those options work with part inheritance.

# The use cases, as I see them

- hide a part
- hide a part's top-level `from` dependency
- hide a part's top-level `after` dependencies
- hide a part's nested `from` dependencies
- hide a part's nested `after` dependencies
- hide specific dependency parts

These use cases can of course be combined for something as complicated as, for example, **hide a part's nested `from` and `after` dependencies as well as its top-level `from` dependency, but not it's top-level `after` dependencies**

# The proposal
Using Joost's [suggestion here](https://discord.com/channels/698854858052075530/944926520282054666/1080169957893087273) as a jumping off point, I think that consolidating all these goals under a single option called `hide` is ideal.

I propose a polymorphic option that relies mainly on 3 strings, `'all'`, `'self'`, or `'nested'`
with these strings you can do the following:

## Basic Hiding
- Hide a part: `{hide: 'self'}`
- Hide a part and all types of dependencies at all levels: `{hide: 'all'}`
- Hide a part's nested dependencies of all types: `{hide: 'nested'}`

_A limitation here is maybe that in basic hiding there isn't something to cover a "hide dependencies of all types at all levels" use case, but our current system doesn't have that option at all, so I'm hoping it's acceptable to have that one happen in the next section_

## Granular hiding
To have more granular control, you can use an object instead, passing `'all'`, `'self'`, or `'nested'` to keys `after` and `from`
- Hide a part's `after` dependencies at all levels: `{hide: {after: 'all'}}`
- Hide only a part's top-level `after` dependencies: `{hide: {after: 'self'}}`
- Hide only a part's nested `after` dependencies: `{hide: {after: 'nested'}`
- Hide a part's `from` dependencies at all levels: `{hide: {from: 'all'}}`
- Hide only a part's top-level `from` dependencies: `{hide: {from: 'self'}}`
- Hide only a part's nested `from` dependencies: `{hide: {from: 'nested'}`

## Combinations
For the above use case of **"hide a part's nested `from` and `after` dependencies as well as its top-level `from` dependency, but not it's top-level `after` dependencies"**, you would use
```js
{
  hide: {
    from: 'all',
    after: 'nested'
  }
}
```

For **"hide dependencies of all types at all levels"**, there's
```js
{
  hide: {
    from: 'all',
    after: 'all'
  }
}
```

## Specific Hiding
To cover the "hide specific dependency parts" use case, we look for one more key in the object, `parts`,  like so: 
```js
{
   hide: {
    parts: ['brian.back']
  }
}

**Comment by woutervdub** [comment]
Thanks for starting this. I have only been pragmatically working with the hiding, and Benjamin helped me quite a bit with that (thanks!), so I feel inadequately prepared for this discussion. One remark I would like to make is this. All of this is looking 'down' the chain of 'inherited' parts. Is it useful to be able to look 'up' too? As in "I do not want this part to be displayed in any design that uses me as a basis, unless they explicitly unhide me".

**Reply by eriese** [reply]
feels risky but exciting. Definitely worth considering, especially if other designers also feel that that'd be more convenient

**Reply by joostdecock** [reply]
> I do not want this part to be displayed in any design that uses me as a basis, unless they explicitly unhide me

Isn't that what (currently) `hide: true` does, and in this proposal `hide: 'self'` would do?

I guess I don't see the use case here, unless it is: *I want to show this part in my own design but I don't want it to show up in other designs* which would raise eyebrows.

We can discuss this on the call. I don't think we need people intimately familiar with how core works under the hood. If anything we should strive for an API that is intuitive to those who don't know or care what happens under the hood. As such, I'd love to hear from all (aspiring) designers.

**Reply by eriese** [reply]
I read it as _I want to show this part in my own design but I don't want it to show up in other designs_, which sounds hinky until I remember that it's more likely for a design to not want any inherited parts to show than vice versa.

And yes, this discussion isn't about how it would work under the hood, but rather about how designers will call it. So y'all are the important demographic, and the question on this proposal is whether it makes sense or feels like a big headache

**Comment by eriese** [comment]
An alternate proposal, more like what Joost was going for in the discord discussion, and as I'm writing it I'm realizing it's probably more inuitive:

```js
{
  hide: {
    self: false,
    from: true,
    after: false,
    nestedFrom: true,
    nestedAfter: true,
    parts: ['brian.back']
  }
}
```

My initial hesitations were that I personally don't like to keep track of that many keys so I wanted an option to do it shorthand, and I thought that it'd be easy to assume that `from` covered both top-level and nested `from` dependencies, but I think that one look at the docs should clear that assumption up quickly.

I definitely think this does more speaking for itself and will therefore be simpler to document. It hasn't needed the lengthy explanation my first proposal did, and that probably makes the extra keystrokes worth it.

**Comment by eriese** [comment]
Ok, based on everyone's feedback from the call yesterday, here's a revised proposal

```js
{ 
  hide: {
    self: false,
    from: true,
    after: false,
    inheritedFrom: true,
    inheritedAfter: 2
  }
}
```
For the "inherited" options if you pass a number, it means "starting at depth" (or more technically "as deep or deeper than"), with 0 being the  depth that the part's direct `after`s are at, 1 being parents, 2 grandparent, and so on. So that here this part's parent `after`s won't be hidden, but any deeper `after`s will be. 

`{inheritedAfter: 1}` is equivalent to `{inheritedAfter: true}`. 
`{inheritedAfter:0}` will be interpreted as `{inheritedAfter: false}`. 

Everyone was clear that "nested" was not intuitive... Is "inherited" better, and if not, what would be? There was also mention of `from` and `after` being non-intuitive language, and I think we should discuss changing it, but while that is the language in place I think this language should match it so that it's clear that you're hiding things that are brought into the dependency chain using those names.

**Comment by eriese** [comment]
Based on [BG3's feedback ](https://discord.com/channels/698854858052075530/1080614314551029890/1082680289144348704), a happy medium approach between my first and the most recent:

```js
{ 
  hide: {
    self: false, // don't hide the part
    after: 1, // hide `after` parts requested by parents and deeper
    from: true // hide `from` parts requested by this part and deeper,
    parts: ['some.part']
  }
}
```

if you wanted to get complicated and do something like 'hide the direct `from` dependency, show the parent `from` dependency, and hide all other `from`s' you could do 
```js
{ 
  hide: {
    from: 2 // hide `from` parts requested by grandparents and deeper,
    parts: ['fromPart']
  }
}
```

Up for discussion is whether we want the part level to be indexed at 0 or 1.

To note:
In all of these proposals, top-level configs should override inherited configs, and if there are conflicts between top-level configs, the stricter one will be what is used

**Reply by joostdecock** [reply]
I am perhaps missing something here, but my takeaway from the contributor call was that a consensus was reached, albeit with the remark that `nested` as a term could perhaps be improved.

**Comment by eriese** [comment]
I am now here with my thinking about how we do this:

```js
{ 
  hide: {
    self: false, // do not hide the part
    from: true, // hide the explicitly included `from` part
    after: false, // do not hide the explicitly included `after` part
    inherited: true, // hide all parts included as `from` or `after` parts by any part in the `from` chain`. Hide all parts included as `from` by any part in the `after` chain
    always: ['randomPart'], // hide these specific parts
    never: ['importantPart'] // never hide these parts
  }
}
 ```

**Comment by joostdecock** [comment]
Implemented. Closing.
